src/common/headers.rs

```diff
@@ -289,6 +295,29 @@ impl Headers {
     pub fn set_accept(&mut self, media_type: MediaType) {
         self.accept = media_type;
     }
+
+    /// Insert a new custom header and value pair into the `HashMap`.
+    pub fn insert_custom_header_line(
+        &mut self,
+        key: String,
+        value: String,
+    ) -> Result<(), RequestError> {
+        if self.custom_map.len() < MAX_CUSTOM_HEADER_LINES {
+            self.custom_map.insert(key, value);
+            return Ok(());
+        }
+        Err(RequestError::CustomHeadersLimitExceeded)
+    }
+
+    /// Returns a custom header value.
+    pub fn get_custom_header_value(&self, key: &str) -> Option<&String> {
+        self.custom_map.get(key)
+    }
+
+    /// Returns the custom header `HashMap`.
+    pub fn custom_map(&self) -> HashMap<String, String> {
+        self.custom_map.clone()
+    }
```

Comment:
Not sure if we should expose both `get_custom_header_value()` and `custom_map()` as part of the public interface. If you provide access to the map, not much point in having another function for getting a single value.

If we do keep this getter, use references instead of clones.

```suggestion
    /// Returns the custom header `HashMap`.
    pub fn custom_map(&self) -> &HashMap<String, String> {
        &self.custom_map
    }
```

Reply:
I guess we need to ask ourselves what are the usecases. I guess users of `struct Headers` don't necessarily know the names/keys of incoming headers, so they need to parse/walk the map anyway and handle custom logic for custom header keys.

In this case, it seems that the `custom_map()` getter is required. We could drop `get_custom_header_value()` as it's simply a wrapper over `HashMap` public `get()` api.

Reply:
The other option would be to add a method which returns an `Iter<>` for the elements of the map, but it just seems convoluted. I like @acatangiu 's approach more.

Reply:
I removed `get_custom_header_value()`, as it does not provide any real value to the `struct Headers` API.